File: backend/__SLUG___node/src/shared/pagination/pagination.ts
import type { Request } from "express";

export type SortDir = "asc" | "desc";

export type FilterOp = "eq" | "contains" | "gt" | "gte" | "lt" | "lte";

export interface FilterParam {
  field: string;
  op: FilterOp;
  value: string;
}

export interface Params {
  page: number;
  pageSize: number;
  sort: string;
  sortDir: SortDir;
  filters: FilterParam[];
}

export interface PagedResponse<T> {
  data: T[];
  total: number;
  page: number;
  pageSize: number;
}

export interface ResolvedFilter {
  column: string;
  op: FilterOp;
  value: string;
}

export interface ResolvedParams {
  page: number;
  pageSize: number;
  sort: string; // DB column
  sortDir: SortDir;
  filters: ResolvedFilter[];
}

/** Thrown when a sort or filter field is not present in the caller's column map. */
export class InvalidFieldError extends Error {
  constructor(public readonly field: string) {
    super(`unknown field: ${field}`);
    this.name = "InvalidFieldError";
  }
}

/** Parses and validates Params from the request body, applying the same defaults/clamps as the Go backend. */
export function decode(req: Request): Params {
  const body = (req.body ?? {}) as Partial<Params>;

  let pageSize = typeof body.pageSize === "number" ? body.pageSize : 25;
  if (pageSize <= 0 || pageSize > 200) {
    pageSize = 25;
  }

  let page = typeof body.page === "number" ? body.page : 0;
  if (page < 0) {
    page = 0;
  }

  const sortDir: SortDir = body.sortDir === "desc" ? "desc" : "asc";

  return {
    page,
    pageSize,
    sort: body.sort ?? "",
    sortDir,
    filters: body.filters ?? [],
  };
}

/** Maps field names to DB columns using the provided columnMap. Throws InvalidFieldError for unknown fields. */
export function resolve(p: Params, columnMap: Record<string, string>): ResolvedParams {
  const resolved: ResolvedParams = {
    page: p.page,
    pageSize: p.pageSize,
    sort: "",
    sortDir: p.sortDir,
    filters: [],
  };

  if (p.sort) {
    const col = columnMap[p.sort];
    if (!col) {
      throw new InvalidFieldError(p.sort);
    }
    resolved.sort = col;
  }

  for (const f of p.filters ?? []) {
    const col = columnMap[f.field];
    if (!col) {
      throw new InvalidFieldError(f.field);
    }
    resolved.filters.push({ column: col, op: f.op, value: f.value });
  }

  return resolved;
}

/** Returns the SQL OFFSET for the current page. */
export function offset(p: ResolvedParams): number {
  return p.page * p.pageSize;
}

File: backend/__SLUG__/src/shared/pagination/pagination.go
package pagination

import (
	"encoding/json"
	"fmt"
	"net/http"
)

type SortDir string

const (
	SortAsc  SortDir = "asc"
	SortDesc SortDir = "desc"
)

type FilterOp string

const (
	OpEq       FilterOp = "eq"
	OpContains FilterOp = "contains"
	OpGt       FilterOp = "gt"
	OpGte      FilterOp = "gte"
	OpLt       FilterOp = "lt"
	OpLte      FilterOp = "lte"
)

type FilterParam struct {
	Field string   `json:"field"`
	Op    FilterOp `json:"op"`
	Value string   `json:"value"`
}

type Params struct {
	Page     int           `json:"page"`
	PageSize int           `json:"pageSize"`
	Sort     string        `json:"sort"`
	SortDir  SortDir       `json:"sortDir"`
	Filters  []FilterParam `json:"filters"`
}

type PagedResponse[T any] struct {
	Data     []T `json:"data"`
	Total    int `json:"total"`
	Page     int `json:"page"`
	PageSize int `json:"pageSize"`
}

// ResolvedFilter holds a filter with its DB column already mapped.
type ResolvedFilter struct {
	Column string
	Op     FilterOp
	Value  string
}

// ResolvedParams holds pagination params with sort field mapped to a DB column.
type ResolvedParams struct {
	Page     int
	PageSize int
	Sort     string // DB column
	SortDir  SortDir
	Filters  []ResolvedFilter
}

// Decode parses and validates a Params from the request body.
func Decode(r *http.Request) (Params, error) {
	var p Params
	p.Page = 0
	p.PageSize = 25
	p.SortDir = SortAsc

	if err := json.NewDecoder(r.Body).Decode(&p); err != nil {
		return p, fmt.Errorf("invalid request body: %w", err)
	}
	if p.PageSize <= 0 || p.PageSize > 200 {
		p.PageSize = 25
	}
	if p.Page < 0 {
		p.Page = 0
	}
	if p.SortDir != SortAsc && p.SortDir != SortDesc {
		p.SortDir = SortAsc
	}
	return p, nil
}

// InvalidFieldError is returned when a sort or filter field is not in the column map.
type InvalidFieldError struct{ Field string }

func (e *InvalidFieldError) Error() string { return "unknown field: " + e.Field }

// Resolve maps field names to DB columns using the provided columnMap.
// Returns an InvalidFieldError if any field is unknown.
func Resolve(p Params, columnMap map[string]string) (ResolvedParams, error) {
	resolved := ResolvedParams{
		Page:     p.Page,
		PageSize: p.PageSize,
		SortDir:  p.SortDir,
	}

	if p.Sort != "" {
		col, ok := columnMap[p.Sort]
		if !ok {
			return resolved, &InvalidFieldError{Field: p.Sort}
		}
		resolved.Sort = col
	}

	for _, f := range p.Filters {
		col, ok := columnMap[f.Field]
		if !ok {
			return resolved, &InvalidFieldError{Field: f.Field}
		}
		resolved.Filters = append(resolved.Filters, ResolvedFilter{
			Column: col,
			Op:     f.Op,
			Value:  f.Value,
		})
	}

	return resolved, nil
}

// Offset returns the SQL OFFSET for the current page.
func (p ResolvedParams) Offset() int {
	return p.Page * p.PageSize
}
